mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/quota/InMemoryCurrentQuotaManager.java

```diff
@@ -74,19 +76,15 @@ public QuotaCountUsage getCurrentMessageCount(QuotaRoot quotaRoot) throws Mailbo
     @Override
     public QuotaSizeUsage getCurrentStorage(QuotaRoot quotaRoot) throws MailboxException {
         try {
-            return quotaCache.get(quotaRoot).size();
+            return quotaCache.get(quotaRoot).get().size();
         } catch (ExecutionException e) {
             throw new MailboxException("Exception caught", e);
         }
     }
 
-    private void updateQuota(QuotaRoot quotaRoot, long count, long size) throws MailboxException {
+    private void updateQuota(QuotaRoot quotaRoot, CurrentQuotas updateQuota, BinaryOperator<CurrentQuotas> quotaFunction) throws MailboxException {
         try {
-            CurrentQuotas currentQuotas = quotaCache.get(quotaRoot);
-            long updatedCount = currentQuotas.count().asLong() + count;
-            long updatedSize = currentQuotas.size().asLong() + size;
-
-            quotaCache.put(quotaRoot, new CurrentQuotas(QuotaCountUsage.count(updatedCount), QuotaSizeUsage.size(updatedSize)));
+            quotaCache.get(quotaRoot).accumulateAndGet(updateQuota, quotaFunction);
```

Comment:
I didn't know `accumulateAndGet`, it's cool.
However, you can probably pass only the BinaryOperator to updateQuota, you don't really need the updateQuota argument, do you?

Reply:
Hmm I need it, `updateQuota` contains the count and size values i need to increase (or decrease) to the entry in the `quotaCache`. Maybe it's the code that is a bit unclear? I guess you can understand it like this, if you cut the logic in two:
```
AtomicReference<CurrentQuotas> currentQuotas = quotaCache.get(quotaRoot);
currentQuotas.accumulateAndGet(updateQuota, quotaFunction);
```

So I do need updateQuota, or I don't know how much I need to increase or decrease my quotas. Do you think it makes sense? Or maybe should I make the code clearer?

Reply:
just pushed a fixup here https://github.com/linagora/james-project/pull/3275/commits/374abb8d1481275457d92d94c3aabddcf7ceeee3

Reply:
Ok I see. Thanks for the help !